packages/datadog_tracking_http_client/lib/src/tracking_http.dart

```diff
@@ -0,0 +1,204 @@
+// ignore_for_file: invalid_use_of_internal_member
+
+import 'dart:async';
+import 'dart:io';
+
+import 'package:datadog_flutter_plugin/datadog_flutter_plugin.dart';
+import 'package:datadog_flutter_plugin/datadog_internal.dart';
+import 'package:flutter/foundation.dart';
+import 'package:http/http.dart' as http;
+import 'package:uuid/uuid.dart';
+
+/// A composable client for use with the `http` package that supports tracking
+/// network requests and sending them to Datadog.
+///
+/// If the RUM feature is enabled, the SDK will send information about RUM
+/// Resources (calling startLoading, stopLoading, and stopLoadingWithErrorInfo)
+/// for all intercepted requests.
+///
+/// The SDK will also create a tracing Span for each 1st-party request, and add
+/// extra HTTP headers to further propagate the trace. The percentage of
+/// resources traced in this way is determined by
+/// [RumConfiguration.tracingSamplingRate].
+///
+/// To specify which hosts are 1st party (and therefore should have tracing
+/// Spans sent), see [DdSdkConfiguration.firstPartyHosts]. You can also set
+/// first party hosts after initialization setting [DatadogSdk.firstPartyHosts]
+///
+/// DatadogClient only modifies calls made through itself, unlike
+/// [DatadogTrackingHttpClient], which overrides all calls made by [HttpClient]
+/// and includes network calls made by Flutter. However, DatadogClient allows
+/// you to compose with other [http.BaseClient] based libraries like
+/// `cupertino_http` and `cronet_http`, which [DatadogTrackingHttpClient] would
+/// miss.
+///
+/// DatadogClient and [DatadogTrackingHttpClient] can be used together if needed,
+/// and will not interfere with each other.
+///
+/// See also [DatadogTrackingHttpClient]
+class DatadogClient extends http.BaseClient {
+  final DatadogSdk datadogSdk;
+  final Uuid _uuid = const Uuid();
+  final http.Client _innerClient;
+
+  DatadogClient({
+    required this.datadogSdk,
+    http.Client? innerClient,
+  }) : _innerClient = innerClient ?? http.Client();
+
+  @override
+  Future<http.StreamedResponse> send(http.BaseRequest request) async {
+    final rum = datadogSdk.rum;
+    // Never track on web
+    if (kIsWeb || rum == null) {
+      return _innerClient.send(request);
+    }
+
+    return _trackingSend(request, rum);
+  }
+
+  Future<http.StreamedResponse> _trackingSend(
+      http.BaseRequest request, DdRum rum) async {
+    String? rumKey;
+    bool isFirstParty = false;
+
+    try {
+      isFirstParty = datadogSdk.isFirstPartyHost(request.url);
+      rumKey = _uuid.v1();
+      bool shouldSample = isFirstParty && rum.shouldSampleTrace();
+      final attributes =
+          _appendRequestHeaders(request, rumKey, isFirstParty, shouldSample);
+
+      final rumHttpMethod = rumMethodFromMethodString(request.method);
+      rum.startResourceLoading(
+          rumKey, rumHttpMethod, request.url.toString(), attributes);
+    } catch (e, st) {
+      datadogSdk.internalLogger.sendToDatadog(
```

Comment:
How often this and another `catch` blocks below can happen? If it can happen just because there is no network connection, won't we spam our telemetry too much?